lfs/api.go

```diff
@@ -0,0 +1,93 @@
+package lfs
+
+import (
+	"io"
+	"sync"
+)
+
+// An abstract interface providing state & resource management for a specific Endpoint across
+// potentially multiple requests
+type ApiContext interface {
+	// Get the endpoint this context was constructed from
+	Endpoint() Endpoint
+	// Close the context & any resources it's using
+	Close() error
+
+	// Download a single object, return reader for data, size and any error
+	// Essentially the same as calling DownloadCheck() then DownloadObject()
+	Download(oid string) (io.ReadCloser, int64, *WrappedError)
+	// Check whether an object is available for download and return an object resource if so
+	DownloadCheck(oid string) (*ObjectResource, *WrappedError)
+	// Download a single object from an already identified resource from DownloadCheck(), return reader for data, size and any error
+	DownloadObject(obj *ObjectResource) (io.ReadCloser, int64, *WrappedError)
+	// Check whether an upload would be accepted for an object and return the resource to use if so
+	UploadCheck(oid string, sz int64) (*ObjectResource, *WrappedError)
+	// Perform the actual upload of an object having identified it will be accepted and the resource to use
+	UploadObject(o *ObjectResource, reader io.Reader) *WrappedError
+	// Perform a batch request for a number of objects to determine what can be uploaded/downloaded
+	Batch(objects []*ObjectResource) ([]*ObjectResource, *WrappedError)
```

Comment:
Honestly, I'd love to reject all other implementations and figure out some way to implement custom ones through commands. Until then, I think we can merge this and move forward with the API design incrementally. We're likely not accepting other implementations any time soon. Each client makes it that much harder to evolve and maintain. 

I'd really like to get rid of `GetApiContext` if we don't need it. That can wait until everything is using the `*All()` batch methods though.